common/functions/src/scalars/hashes/sha2hash.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2020 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+use common_datavalues::columns::DataColumn;
+use common_datavalues::prelude::*;
+use common_datavalues::DataSchema;
+use common_datavalues::DataType;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use sha2::Digest;
+
+use crate::scalars::function_factory::FunctionDescription;
+use crate::scalars::function_factory::FunctionFeatures;
+use crate::scalars::Function;
+
+#[derive(Clone)]
+pub struct Sha2HashFunction {
+    display_name: String,
+}
+
+impl Sha2HashFunction {
+    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
+        Ok(Box::new(Sha2HashFunction {
+            display_name: display_name.to_string(),
+        }))
+    }
+
+    pub fn desc() -> FunctionDescription {
+        FunctionDescription::creator(Box::new(Self::try_create))
+            .features(FunctionFeatures::default().deterministic())
+    }
+}
+
+impl Function for Sha2HashFunction {
+    fn name(&self) -> &str {
+        &*self.display_name
+    }
+
+    fn num_arguments(&self) -> usize {
+        2
+    }
+
+    fn return_type(&self, args: &[DataType]) -> Result<DataType> {
+        if args[0] == DataType::String
+            && matches!(
+                args[1],
+                DataType::UInt8 | DataType::UInt16 | DataType::UInt32
+            )
+        {
+            Ok(DataType::String)
+        } else {
+            Err(ErrorCode::IllegalDataType(format!(
+                "Expected string and numeric type, but got {}",
+                args[0]
+            )))
+        }
+    }
+
+    fn nullable(&self, _input_schema: &DataSchema) -> Result<bool> {
+        Ok(false)
+    }
+
+    fn eval(&self, columns: &DataColumnsWithField, _input_rows: usize) -> Result<DataColumn> {
+        let i_series = columns[0]
+            .column()
+            .to_minimal_array()?
+            .cast_with_type(&DataType::String)?;
+
+        let l_series = columns[1]
+            .column()
+            .to_minimal_array()?
+            .cast_with_type(&DataType::UInt16)?;
+
+        let opt_iter = i_series
+            .string()?
+            .into_no_null_iter()
+            .zip(l_series.u16()?.into_no_null_iter())
```

Comment:
I'm afraid `i_series` and `l_series` may not have the same size.

Eg:
`select sha2( toString(number),  256   )  from numbers(10)`


`l_series` is more likely to be a scalar value DataValue::UInt16, we can have a specific path for this case, thus we can avoid a dynamic `match l` inside the loop to improve the performance.

Reply:
You are right. Changed code to handle scalar case, e.g: `select sha2( toString(number), 256 ) from numbers(10)` for length of hash2.

Reply:
Great, now it's better yet it's not perfect.

```
 let hash = |i: &[u8], l: &u16| match  ...;
 loop {
      hash(x, y)
 }
```

This did not fit in cpu cache.

Better to use 

```
match l {
    a =>  loop {},
    b => loop {},
     ...
}

```

for the constant case.

But it needs much more duplicated codes.

Reply:
Wow, seems `into_no_null_iter()` did not consider null value case.